src/instancemanager.jl

```diff
@@ -0,0 +1,330 @@
+    m.state = AttachedSolver
+    # MOI does not define the type of index_map, so we have to copy it into a
+    # concrete container.
+    m.instancetosolvermap = IndexMap()
+    for k in keys(index_map)
+        m.instancetosolvermap[k] = index_map[k]
+    end
+    return
+end
+
+# TODO: We need to provide more detail about what causes `cancopy` to return false,
+# e.g., an error message saying which constraint type isn't supported.
+"""
+    canattachsolver(m::InstanceManager)
+
+Returns `true` if the solver reports that it `cancopy` from the instance.
+`false` typically means that the solver does not support the problem type
+stored in the instance. Can be called only from the `EmptySolver` state.
+"""
+function canattachsolver(m::InstanceManager)
+    @assert m.state == EmptySolver
+    return MOI.cancopy(m.solver, m.instance)
+end
+
+# Optimizing and adding/modifying constraints and variables.
+
+function MOI.optimize!(m::InstanceManager)
+    if m.mode == Automatic && m.state == EmptySolver
+        attachsolver!(m)
+    end
+    # TODO: better error message if no solver is set
+    @assert m.state == AttachedSolver
+    MOI.optimize!(m.solver)
+end
+
+function MOI.addvariable!(m::InstanceManager)
+    # TODO: This is currently broken for automatic mode if the solver doesn't
+    # support addvariable!. We probably need a "canaddvariable".
+    vindex = MOI.addvariable!(m.instance)
+    if m.state == AttachedSolver
+        vindex_solver = MOI.addvariable!(m.solver)
+        m.instancetosolvermap[vindex] = vindex_solver
+    end
+    return vindex
+end
+
+function MOI.addvariables!(m::InstanceManager, n)
+    # TODO: Same note as above.
+    vindices = MOI.addvariables!(m.instance, n)
+    if m.state == AttachedSolver
+        vindices_solver = MOI.addvariables!(m.solver, n)
+        for (vindex, vindex_solver) in zip(vindices, vindices_solver)
+            m.instancetosolvermap[vindex] = vindex_solver
+        end
+    end
+    return vindices
+end
+
+function MOI.canaddconstraint(m::InstanceManager, func::MOI.AbstractFunction, set::MOI.AbstractSet)
+    MOI.canaddconstraint(m.instance, func, set) || return false
+    if m.state == AttachedSolver && m.mode == Manual
+        MOI.canaddconstraint(m.solver, func, set) || return false
+    end
+    return true
+end
+
+function MOI.addconstraint!(m::InstanceManager, func::MOI.AbstractFunction, set::MOI.AbstractSet)
+    # The canaddconstraint checks should catch most issues, but if an
+    # addconstraint! call fails then the instance and the solver may no longer
+    # be in sync.
+    if m.mode == Automatic && !MOI.canaddconstraint(m.solver, func, set)
+        resetsolver!(m)
+    end
+    @assert MOI.canaddconstraint(m, func, set)
+    cindex = MOI.addconstraint!(m.instance, func, set)
+    if m.state == AttachedSolver
+        cindex_solver = MOI.addconstraint!(m.solver, func, set)
+        m.instancetosolvermap[cindex] = cindex_solver
+    end
+    return cindex
+end
```

Comment:
If you try adding to solver before adding to instance there is less chance of losing sync here

Reply:
```julia
function MOI.addconstraint!(m::InstanceManager, func::MOI.AbstractFunction, set::MOI.AbstractSet)
     # The canaddconstraint checks should catch most issues, but if an
     # addconstraint! call fails then the instance and the solver may no longer
     # be in sync.
     if m.mode == Automatic && !MOI.canaddconstraint(m.solver, func, set)
         resetsolver!(m)
     end
     @assert MOI.canaddconstraint(m, func, set)
     if m.state == AttachedSolver
         cindex_solver = MOI.addconstraint!(m.solver, func, set)
     end
     cindex = MOI.addconstraint!(m.instance, func, set)
     if m.state == AttachedSolver
         m.instancetosolvermap[cindex] = cindex_solver
     end
     return cindex
 end
```

Reply:
What do you mean ? You are considering the case where the solver throws an error ?

Reply:
Yes, but given that we haven't defined any way for solvers to throw an error, I'll keep the slightly simpler solution for now. We can revisit this if it becomes an issue.